Implements Series.split_into/3

ryancurtin opened the pull request:
Implements the `Series.split_into/3` function, which closes #871.  String columns can be split into a struct with multiple fields.

josevalim commented on lib/explorer/backend/lazy_series.ex:
```diff
@@ -1053,6 +1054,13 @@ defmodule Explorer.Backend.LazySeries do
+  def split_into(series, by, fields) do
+    data = new(:split_into, [lazy_series!(series), by, fields], :string)
+
+    Backend.Series.new(data, {:list, :struct})
```
I think this should be something like `{:struct, Enum.map(fields, &{&1, :string})}`

josevalim commented on lib/explorer/series.ex:
```diff
@@ -5651,6 +5651,31 @@ defmodule Explorer.Series do
     do: dtype_error("split/2", dtype, [:string])
 
+  @doc """
+  Split a string Series into a struct with fields determined by the list of
+  field names provided.  The length of the field names list determines how
+  many fields the resulting struct will have.  If the string cannot be split
+  into that many separate strings, null values will be provided for the
+  remaining fields.
```
```suggestion
  Split a string Series into a struct of string `fields`.
  
  The length of the field names list determines how many times the
  string will be split at most. If the string cannot be split into that
  many separate strings, null values will be provided for the
  remaining fields.
```

The first sentence in docs is always a short summary. :)

josevalim commented:
This is looking great so far @ryancurtin! The last step is to implement the feature for the lazy backend as well and add a test. The lazy backend is the one used inside DF.mutate and friends. You can see other PRs, such as this one, for examples: https://github.com/elixir-explorer/explorer/pull/859

ryancurtin commented:
@josevalim - Ok, I've re-requested review since I've now implemented the lazy backend and added some tests using a DataFrame.  Thank you again for your help!

lkarthee commented on native/explorer/src/series.rs:
```diff
@@ -1567,6 +1567,26 @@ pub fn s_split(s1: ExSeries, by: &str) -> Result<ExSeries, ExplorerError> {
     Ok(ExSeries::new(s2))
 }
 
+#[rustler::nif(schedule = "DirtyCpu")]
```
Series can be converted into lazy frame and the column/series can be manipulated using expression. That way we can avoid using ChunkedArrays.

```rust
let s2 = s1
        .clone_inner()
        .into_frame()
        .lazy()
        .select([col(s1.name())
            .str()
        // ops - mirroring expr_split_into
            .alias("res")])
        .collect()?
        .column("res")?
        .clone();
Ok(ExSeries::new(s2))
```

josevalim commented on lib/explorer/series.ex:
```diff
@@ -5651,6 +5651,32 @@ defmodule Explorer.Series do
+  """
+  @doc type: :string_wise
+  @spec split_into(Series.t(), String.t(), list(String.t() | atom())) :: Series.t()
+  def split_into(%Series{dtype: :string} = series, by, fields) when is_binary(by),
+    do: apply_series(series, :split_into, [by, fields])
+
+  def split_into(%Series{dtype: dtype}, _by, _fields),
+    do: dtype_error("split_into/3", dtype, [:string])
```
```suggestion
  def split_into(%Series{dtype: :string} = series, by, [_ | _] = fields) when is_binary(by),
    do: apply_series(series, :split_into, [by, Enum.map(fields, &to_string/1)])

  def split_into(%Series{dtype: dtype}, by, [_ | _] = _fields)  when is_binary(by),
    do: dtype_error("split_into/3", dtype, [:string])
```

josevalim replied:
Let's normalize fields to string before and make sure it is a non-empty list.

ryancurtin-wk replied:
@josevalim - The underlying polars `splitn` function takes a list of arbitrary length.  Do you think it makes sense to limit the explorer API to only use a list of two fields?  Alternatively, I could put a guard to ensure that `fields` is a list so that users can split the column however many times necessary.

josevalim commented:
Three remaining comments and we will ship it :D

josevalim commented on lib/explorer/series.ex:
```diff
@@ -5651,6 +5651,32 @@ defmodule Explorer.Series do
     do: dtype_error("split/2", dtype, [:string])
 
+  @doc """
+  Split a string Series into a struct of string `fields`.
+  
+  The length of the field names list determines how many times the
```
```suggestion
  Split a string Series into a struct of string `fields`.

  The length of the field names list determines how many times the
```

josevalim commented:
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
